src/adapter/src/coord/cluster_scheduling.rs

```diff
@@ -0,0 +1,253 @@
+use mz_sql::catalog::CatalogCluster;
+use mz_sql_parser::ast::ClusterScheduleOptionValue;
+use std::time::Instant;
+use timely::progress::Antichain;
+use tracing::{debug, warn};
+
+const POLICIES: &[&str] = &[REFRESH_POLICY_NAME];
+
+const REFRESH_POLICY_NAME: &str = "refresh";
+
+impl Coordinator {
+    #[mz_ore::instrument(level = "debug")]
+    /// Call each scheduling policy.
+    pub(crate) async fn check_scheduling_policies(&mut self) {
+        // (So far, we have only this one policy.)
+        self.check_refresh_policy();
+    }
+
+    /// Runs the `SCHEDULE = ON REFRESH` cluster scheduling policy, which makes cluster On/Off
+    /// decisions based on REFRESH materialized view write frontiers and the current time (the local
+    /// oracle read ts), and sends `Message::SchedulingDecisions` with these decisions.
+    /// (Queries the timestamp oracle on a background task.)
+    fn check_refresh_policy(&mut self) {
+        let start_time = Instant::now();
+
+        // Collect the smallest REFRESH MV write frontiers per cluster.
+        let mut min_refresh_mv_write_frontiers = Vec::new();
+        for cluster in self.catalog().clusters() {
+            if let ClusterVariant::Managed(ref config) = cluster.config.variant {
+                match config.schedule {
+                    ClusterScheduleOptionValue::Manual => {
+                        // Nothing to do, user manages this cluster manually.
+                    }
+                    ClusterScheduleOptionValue::Refresh => {
+                        let refresh_mv_write_frontiers = cluster
+                            .bound_objects()
+                            .iter()
+                            .filter_map(|id| {
+                                if let CatalogItem::MaterializedView(mv) =
+                                    self.catalog().get_entry(id).item()
+                                {
+                                    if mv.refresh_schedule.is_some() {
+                                        Some(&self
+                                            .controller
+                                            .storage
+                                            .collection(*id)
+                                            .expect("the storage controller should know about MVs that exist in the catalog")
+                                            .write_frontier)
+                                    } else {
+                                        None
+                                    }
+                                } else {
+                                    None
+                                }
+                            })
+                            .collect_vec();
+                        debug!(
+                            "check_refresh_policy: cluster_id: {}, refresh_mv_write_frontiers: {:?}",
+                            cluster.id,
+                            refresh_mv_write_frontiers,
+                        );
```

Comment:
just a tip, you can rewrite `tracing` statements like
```
debug!(cluster.id, ?refresh_mv_write_frontiers, "check_refresh_policy")
```
and it will produce a trace like
```
cluster.id = <val>, refresh_mv_write_frontiers = <val>, "check_refresh_policy"
```

Reply:
Thanks, I've simplified some of the debug prints.